dlib/cuda/tensor_tools.h

```diff
@@ -158,28 +158,66 @@ namespace dlib { namespace tt
 
+    enum gemm_mode { CHANNEL_WISE = 0, PLANE_WISE = 1 };
+
     void gemm (
         float beta,
         tensor& dest,
         float alpha,
         const tensor& lhs,
         bool trans_lhs,
         const tensor& rhs,
-        bool trans_rhs
+        bool trans_rhs,
+        gemm_mode g_mode = CHANNEL_WISE
     );
     /*!
         requires
             - dest does not alias the memory of lhs or rhs
             - The dimensions of lhs and rhs must be compatible for matrix multiplication.
-              In particular:
+            The specific requirements depend on the gemm_mode:
+
+            For CHANNEL_WISE mode (default):
                 - Let L == trans_lhs ? trans(mat(lhs)) : mat(lhs)
                 - Let R == trans_rhs ? trans(mat(rhs)) : mat(rhs)
                 - Let D == mat(dest)
                 - D.nr() == L.nr() && D.nc() == R.nc()
-                  (i.e. dest must be preallocated and have the correct output dimensions)
+                (i.e. dest must be preallocated and have the correct output dimensions)
                 - L.nc() == R.nr()
-        ensures
-            - performs: dest = alpha*L*R + beta*mat(dest)
+
+            For PLANE_WISE mode:
+                - lhs.num_samples() == rhs.num_samples() && lhs.k() == rhs.k()
+                - If !trans_lhs && !trans_rhs:
+                    lhs.nc() == rhs.nr()
+                    dest.nr() == lhs.nr() && dest.nc() == rhs.nc()
+                - If trans_lhs && !trans_rhs:
+                    lhs.nr() == rhs.nr()
+                    dest.nr() == lhs.nc() && dest.nc() == rhs.nc()
+                - If !trans_lhs && trans_rhs:
+                    lhs.nc() == rhs.nc()
+                    dest.nr() == lhs.nr() && dest.nc() == rhs.nr()
+                - If trans_lhs && trans_rhs:
+                    lhs.nr() == rhs.nc()
+                    dest.nr() == lhs.nc() && dest.nc() == rhs.nr()
+
+        ensures
+            - Performs matrix multiplication based on the specified gemm_mode:
+
+            For CHANNEL_WISE mode:
+                - performs: dest = alpha*L*R + beta*mat(dest)
+                Where L, R, and D are as defined above.
+
+            For PLANE_WISE mode:
+                - Performs matrix multiplication for each corresponding 2D plane (nr x nc)
+                in lhs and rhs across all samples and channels.
+                - The operation is equivalent to performing the following for each sample
+                and channel:
+                    dest[s][k] = alpha * (lhs[s][k] * rhs[s][k]) + beta * dest[s][k]
+                Where [s][k] represents the 2D plane for sample s and channel k.
+
+        note
+            - The PLANE_WISE mode is particularly useful for operations like attention
+            mechanisms in neural networks, where you want to perform matrix multiplications
+            on 2D planes of 4D tensors while preserving the sample and channel dimensions.
```

Comment:
```suggestion
            - The PLANE_WISE mode is particularly useful for operations like attention
              mechanisms in neural networks, where you want to perform matrix multiplications
              on 2D planes of 4D tensors while preserving the sample and channel dimensions.
```
Here and elsewhere, please make sure the comments have indenting that lines up like how you see it in other parts of the library.

Reply:
It might also be related to the text editor I'm using. I'll keep an eye on that for future modifications. In any case, thank you for this update.

Reply:
I've changed it a bit to match the other notes